src/keys/bip39.rs

```diff
@@ -5,34 +5,216 @@
 
 // https://doc.rust-lang.org/std/primitive.str.html
 // "String slices are always valid UTF-8."
-type Mnemonic = str;
-type Passphrase = str;
-type Seed = [u8; 64];
 
 use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+use core::convert::TryFrom;
+use core::ops::Deref;
 
-use unicode_normalization::UnicodeNormalization;
+use unicode_normalization::{is_nfkd, UnicodeNormalization};
+use zeroize::{Zeroize, ZeroizeOnDrop, Zeroizing};
 
-pub fn mnemonic_to_seed(m: &Mnemonic, p: &Passphrase, s: &mut Seed) {
-    let m = m.chars().nfkd().collect::<String>();
+/// BIP39 coded error.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    /// Mnemonic entropy amount is invalid (should be 128 or 160 or 192 or 224 or 256 bits)
+    InvalidEntropyCount(usize),
+    /// Mnemonic contains a word not present in word list
+    NoSuchWord(String),
+    /// Mnemonic corrupted, checksum mismatch
+    ChecksumMismatch,
+    /// Mnemonic is not in NFKD form
+    UnnormalizedMnemonic,
+    /// Passphrase is not in NFKD form
+    UnnormalizedPassphrase,
+    /// Word list contains unnormalized word or word with a separator
+    BadWordlistWord(String),
+    /// Word list contains duplicate words
+    UnsortedWordlist,
+    /// Separator is not in NFKD form
+    BadSeparator,
+}
+
+/// Reference to a normalized (unicode NFKD) mnemonic.
+#[derive(Clone, Copy)]
+pub struct MnemonicRef<'a>(&'a str);
```

Comment:
Can you impl `ToOwned` here and `Borrow` for Mnemonic?

Reply:
Same with Password

Reply:
I tried and failed. The major problem is with `ToOwned` which is already implemented for `T: Clone`. With `Borrow` there's generic lifetime argument to `MnemonicRef` which we need to know in advance. And `borrow` need to return reference to `MnemonicRef` not value.